Update schema upgrade docs (#13)
pingidentity/pingauthorize-contrib: sql/postgresql/README.md

> # PingAuthorize Policy Editor PostgreSQL Scripts (9.0 - 9.1)

Reviewer: > **Warning** 
The following instructions apply only to PingAuthorize versions 9.0.x.x and 9.1.x.x. Earlier versions did not support PostgreSQL, and later versions require the `db-cli` tool to initialize and upgrade PostgreSQL databases.

psql --dbname=my_pap_db --file="<pingauthorize-contrib-root>/sql/postgresql/init-policy-database.sql" >/dev/null
```
You should now have an initialized policy database. For instructions on configuring the Policy Editor to connect to
your PostgreSQL instance, see [Installing the PingAuthorize Policy Editor manually](https://docs.pingidentity.com/csh?Product=paz-latest&context=paz_install_pe_noninteractive).

### Upgrading an existing policy database to a newer schema
1. Verify the current version of your policy database using one of the following methods.

**Optional** - Access `my_pap_db` and query the `TAG` column of `my_pap_db.databasechangelog`

```

> psql -U postgres -d my_pap_db

Reviewer 2: Another small nit here, but the prerequisites list that the admin should already be signed in as an authenticated PostgreSQL superuser.

```

```
SELECT TAG FROM DATABASECHANGELOG WHERE TAG IS NOT NULL ORDER BY TAG DESC LIMIT 1;
```

**Optional** - Follow the [steps to non-interactively set up](https://docs.pingidentity.com/csh?Product=paz-latest&context=paz_install_pe_noninteractive)

>    the target version of the Policy Editor. If your schema is out of date, you will be prompted to run the intermediate upgrade scripts sequentially.

Reviewer 2: I think it would be worthwhile to add a warning to clarify that `setup` is intended to be run against a clean install of the Policy Editor. The example command below would not work as-is if there are existing `configuration.yml` or certificate keystores/truststores.  It's not the best user experience, but currently this approach sets up a separate Policy Editor install (as opposed to an in-place upgrade) for the sole purpose of identifying the upgrade scripts.

Can you add some clarifying content explaining that during `start-server`, the application checks if the database schema requires upgrades, and if so, prints the locations of the upgrade scripts, and that the `setup` command below is provided if the customer wishes to avoid affecting their existing installation files?

```
bin/setup demo \
  --adminUsername admin \

>   --dbConnectionString "jdbc:postgresql://<host>:<postgresql_port>/my_pap_db" \

Reviewer 2: This is a nit, but the PostgreSQL host does not necessarily need to be the same as the server host.

  --dbAppUsername "<postgres user>" \
  --generateSelfSignedCertificate \
  --decisionPointSharedSecret pingauthorize \
  --hostname <host> \
  --port <pap port> \
  --adminPort <admin port> \
  --licenseKeyFile <license>
```

_Setup completed successfully_

_Please start the server by running bin/start-server_

```
<PingAuthorize-PAP>bin/start-server
```

_The policy database at 'jdbc:postgresql://localhost:5432/my_pap_db' is older than this version of PingAuthorize (9.1.0.0). Please use the following scripts to upgrade the policy database before running start-server again:_

_https://github.com/pingidentity/pingauthorize-contrib/blob/main/sql/postgresql/9.1-GA.sql_

> _https://github.com/pingidentity/pingauthorize-contrib/blob/main/sql/postgresql/9.1-EA.sql_

Reviewer 2: I think the intent was for the URLs to be printed in the order that they should be applied, but here it looks like they are in the reverse order (I would expect the "EA" script to be applied before the "GA" script).

This could be a defect in the `setup` tool. Would you check that and create an internal ticket to address it if that's the case?  As a strawman, I could see it being very helpful to order them correctly and also prefix them with an index to remove any ambiguity.

Author: It appears to be a defect in `setup` as this is the order in which they were printed out. Thanks for catching that.

I will follow up and make sure that is resolved.

psql --dbname=my_pap_db --file="<pingauthorize-contrib-root>/sql/postgresql/init-policy-database.sql" >/dev/null
```
You should now have an initialized policy database. For instructions on configuring the Policy Editor to connect to
your PostgreSQL instance, see [Installing the PingAuthorize Policy Editor manually](https://docs.pingidentity.com/csh?Product=paz-latest&context=paz_install_pe_noninteractive).

### Upgrading an existing policy database to a newer schema
1. Verify the current version of your policy database using one of the following methods.

**Optional** - Access `my_pap_db` and query the `TAG` column of `my_pap_db.databasechangelog`

```
psql --dbname=my_pap_db
```

```

> SELECT TAG FROM DATABASECHANGELOG WHERE TAG IS NOT NULL ORDER BY TAG DESC LIMIT 1;

Reviewer 2: Thinking about this a little bit more, I think we should just remove this content and only document the `start-server` method. This approach requires customers to know about the `DATABASECHANGELOG` table, and I think that `ORDER BY TAG DESC LIMIT 1` breaks once we start moving to `10.0`.

You should now have an initialized policy database. For instructions on configuring the Policy Editor to connect to

> your PostgreSQL instance, see "[Installing the PingAuthorize Policy Editor manually](https://docs.pingidentity.com/csh?Product=paz-latest&context=paz_install_pe_noninteractive)".

Reviewer: ```suggestion
your PostgreSQL instance, see [Installing the PingAuthorize Policy Editor manually](https://docs.pingidentity.com/csh?Product=paz-latest&context=paz_install_pe_noninteractive).
```

### Upgrading an existing policy database to a newer schema

> 1. Verify the current version of your policy database. Below are a couple of easy ways to do this.

Reviewer: ```suggestion
1. Verify the current version of your policy database using one of the following methods.
```

**Optional** - Access `my_pap_db` and query the `TAG` column of `my_pap_db.databasechangelog`

`psql -U postgres -d my_pap_db`

`SELECT TAG FROM DATABASECHANGELOG WHERE TAG IS NOT NULL ORDER BY TAG DESC LIMIT 1;`

> **Optional** - Follow the "[Steps to non-interactively set up](https://docs.pingidentity.com/bundle/pingauthorize-91/page/hnc1585320115289.html#hnc1585320115289)"

Reviewer: ```suggestion
**Optional** - Follow the [Steps to non-interactively set up](https://docs.pingidentity.com/bundle/pingauthorize-91/page/hnc1585320115289.html#hnc1585320115289)
```

   the target version of the Policy Editor. If your schema is out of date, you will be prompted to run the intermediate upgrade scripts sequentially.

`bin/setup demo \
  --adminUsername admin \
  --dbConnectionString "jdbc:postgresql://<host>:<postgresql_port>/my_pap_db" \
  --dbAppUsername "<postgres user>" \
  --generateSelfSignedCertificate \
  --decisionPointSharedSecret pingauthorize \
  --hostname <host> \
  --port <pap port> \
  --adminPort <admin port> \
  --licenseKeyFile <license>`

_Setup completed successfully_

_Please start the server by running bin/start-server_

`<PingAuthorize-PAP>bin/start-server`

_The policy database at 'jdbc:postgresql://localhost:5432/my_pap_db' is older than this version of PingAuthorize (9.1.0.0). Please use the following scripts to upgrade the policy database before running start-server again:_

_https://github.com/pingidentity/pingauthorize-contrib/blob/main/sql/postgresql/9.1-GA.sql_

_https://github.com/pingidentity/pingauthorize-contrib/blob/main/sql/postgresql/9.1-EA.sql_

> 2. Download and apply the upgrade scripts between your current version and the target version.

Reviewer: ```suggestion
2. Download and apply the upgrade scripts for the policy database versions between your current version and the target version.
```

> For example - upgrading from `9.0-GA` to `9.1-GA`. Between these versions, both the `9.1-EA`, and `9.1-GA` upgrade scripts must be applied.

Reviewer: ```suggestion
For example, in order to upgrade from `9.0-GA` to `9.1-GA`, both the `9.1-EA`, and `9.1-GA` upgrade scripts must be applied.
```

`wget https://raw.githubusercontent.com/pingidentity/pingauthorize-contrib/main/sql/postgresql/9.1-EA.sql`

`psql --dbname=my_pap_db --file=9.1-EA.sql >/dev/null`

`wget https://raw.githubusercontent.com/pingidentity/pingauthorize-contrib/main/sql/postgresql/9.1-GA.sql`

`psql --dbname=my_pap_db --file=9.1-GA.sql >/dev/null`

> At this point the upgrade is complete.

Reviewer: ```suggestion
After you apply the necessary upgrade scripts, the policy database upgrade is complete.
```
